### DISCUSSION TITLE
Is it possible to associate different colors with different display text/link copy formats?

### DISCUSSION THREAD

**Discussion by crackrepacktorrent** [RyotaUshio/obsidian-pdf-plus#145] [Q&A] [answered]
I like to incorportate multiple different styles of citations when going through a PDF. Is this already possible?

**Comment by N3C2L** [comment]
Yes, available in the PDF++ settings.

Try to update PDF++ and follow the link to the PDF++ options here that @RyotaUshio provided in his answer:

https://github.com/RyotaUshio/obsidian-pdf-plus/discussions/135#discussioncomment-8960268

Right clicking in the PDF viewer after text highlighting to choose your new created color / link / display formats you have created in the PDF++ Settings.

**Reply by crackrepacktorrent** [reply]
I don't think this is a solution. This lets me change each of the default link copy formats, but it doesn't let me assign them to a specific color. What i would like to do is, for each of the colors on the PDF++ top bar, have a different display text/link copy format associated with them.

**Reply by crackrepacktorrent** [reply]
Bumping this!

**Comment by RyotaUshio** [comment, marked as answer]
It would be nice to have macros, i.e. specific combinations of colors, link copy formats, and display text formats. But currently, PDF++ doesn't support it as a built-in feature.

A possible workaround is tweaking JavaScript templates. For example:

1. Use the following as a Templater startup script. It defines a global function named `setLinkColor`. For example, `setLinkColor('file.pdf#page=1&selection=0,1,2,3&color=whatever-color', 'red')` becomes `'file.pdf#page=1&selection=0,1,2,3&color=red'`.
   
   This function can be called from within `{{  }}` in link copy format templates.

```js
<%*
window.setLinkColor = (linktext, color) => {
	if (!color) {
		throw Error('Color is not specified');
	}
	
	let [path, subpath] = linktext.split('#');
	if (!subpath) {
		throw Error('The given link text does not have a subpath');
	}
	
	const params = new URLSearchParams(subpath);
	params.set('color', color);
	subpath = Array.from(params.entries())
		.map(([key, value]) => key + '=' + value)
		.join('&');
	return path + '#' + subpath;
}
%>
```

2. Define the following link copy formats:
  - "Example 1": 
    ```
    [[{{ setLinkColor(linktext, 'yellow') }}|{{file.basename}}, page {{pageLabel}}]]
    ```
  - "Example 2"
    ```
    [[{{ setLinkColor(linktext, 'red') }}|{{text}}]]
    ```
    
3. Then, whenever you trigger the "Example 1" link copy format, you will get the same result as selecting
  - the yellow color,
  - the display text format `{{file.basename}}, page {{pageLabel}}`, and
  - the link copy format `{{linkWithDisplay}}`
  regardless of the actual color and display text format you selected.
  
Here I assumed that you are enabling the "use [[wikilinks]]" option in Obsidian's files & links settings. If not, you will need to modify the examples above accordingly. For example, "Example 1" should be rewritten as

```
[{{file.basename}}, page {{pageLabel}}]({{ setLinkColor(linktext, 'yellow') }})
```

**Reply by RyotaUshio** [reply]
FYI, I'm considering implementing macros as a part of #159.
